src/ops/reduction.rs

```diff
@@ -0,0 +1,298 @@
+use crate::array::Array;
+use crate::error::OperationError;
+use crate::stream::StreamOrDevice;
+use mlx_macros::default_device;
+
+impl Array {
+    /// An `and` reduction over the given axes.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use mlx::Array;
+    /// let a = Array::from_slice(&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], &[3, 4]);
+    /// let mut b = a.all_axes(&[0], None);
+    ///
+    /// b.eval();
+    /// let results: &[bool] = b.as_slice();
+    /// // results == [false, true, true, true]
+    /// ```
+    ///
+    /// # Params
+    ///
+    /// - axes: The axes to reduce over
+    /// - keep_dims: Whether to keep the reduced dimensions -- defaults to false if not provided
+    /// - stream: The stream to execute the operation on
+    #[default_device]
+    pub fn all_axes_device(
+        &self,
+        axes: &[i32],
```

Comment:
The python api allows `axes` to be optional. Should this be optional too?

Reply:
python allows calling `all` with either an int, a vec of ints or nothing -- in Rust I've split these into three methods: `all_axes (takes array)`, `all_axis (takes one)` or `all` which takes none.

I'm not sure there's a good way in Rust to denote these three different scenarios in one.

Reply:
I think in mlx c++ code the integer axis is wrapped into `std::vector<int>` and then call the array axes variant. So maybe `fn all(&self, axes: Option<&[i32]>, keep_dims: Option<bool>, s: StreamOrDevice)` is fine? The user would just need to wrap a single axis into a single element slice themselves?

Reply:
Fine with me! Makes things simpler. Will make the change.